superdesk/http_proxy.py

```diff
@@ -0,0 +1,160 @@
+from flask import Flask, request, current_app, Response as FlaskResponse, make_response
+
+
+class HTTPProxy:
+    """A class for proxying HTTP requests to an external service.
+
+    Attributes
+    ----------
+    endpoint_name : str
+        The endpoint name used when registering the proxy with Flask (used in ``url_for``)
+    service_url : str
+        The external URL used to proxy the requests to
+    base_url : int
+        The internal URL used to proxy requests from
+    http_methods: list[str]
+        list of HTTP methods allowed for this proxy
+    auth: bool
+        If ``True`` (the default), then this proxy is available only to authenticated users
+    use_cors: bool
+        if ``True`` (the default), then CORS is enabled on this proxy
+
+    To register a new proxy, use the ``register_http_proxy`` method with an instance of a HTTPProxy class
+
+    Example:
+    ::
+        register_http_proxy(
+            application,
+            HTTPProxy(
+                endpoint_name="belga_ai_proxy",
+                service_url="belga/ai",
+                base_url="http://localhost:5001",
+            )
+        )
+
+    Note: The ``service_url`` is prefixed with the ``URL_PREFIX`` config (defaults to ``api``).
+
+    Example Requests:
+    ::
+        http://localhost:5000/api/belga/ai                 -> http://localhost:5001
+        http://localhost:5000/api/belga/ai/articles/123456 -> http://localhost:5001/articles/123456
+        url_for("belga_ai_proxy")                          -> /api/belga/ai
+        url_for("belga_ai_proxy", _external=True)          -> http://localhost:5000/api/belga/ai
+    """
+
+    endpoint_name: str
+    service_url: str
+    base_url: str
+    http_methods: List[str]
+    auth: bool
+    use_cors: bool
+    session: requests.Session
+
+    def __init__(
+        self,
+        endpoint_name: str,
+        base_url: str,
+        service_url: str,
+        http_methods: Optional[List[str]] = None,
+        auth: bool = True,
+        use_cors: bool = True,
+    ):
+        self.endpoint_name = endpoint_name
+        self.base_url = base_url.lstrip("/").rstrip("/")
+        self.service_url = service_url
+        self.http_methods = http_methods or ["OPTIONS", "GET", "POST", "PATCH", "DELETE"]
+        self.auth = auth
+        self.use_cors = use_cors
+        self.session = requests.Session()
+
+    def get_base_url(self, app: Optional[Flask] = None) -> str:
+        """Returns the base URL route used when registering the proxy with Flask"""
+
+        url_prefix = ((app or current_app).config["URL_PREFIX"]).lstrip("/")
+        return f"/{url_prefix}/{self.base_url}"
+
+    def process_request(self, path: str) -> FlaskResponse:
+        """The main function used for processing requests from the client"""
+
+        self.authenticate()
+        if request.method == "OPTIONS":
+            return self.construct_cors_response()
+
+        result = self.session.request(**self.get_proxy_request_kwargs())
```

Comment:
might be useful to have this in a separate method so we can override this in a subclass but would still have that auth/cors handling there